File: Data Structures & Algorithms/combination-target-sum/submission-2.py
class Solution:
    def combinationSum(self, nums: List[int], target: int) -> List[List[int]]:
        res = []

        def helper(curr, total, index):
            if total == target:
                res.append(curr.copy())
                return
            elif total > target or index == len(nums):
                return
            
            num = nums[index]

            curr.append(num)
            helper(curr, total + num, index)
            curr.pop()
            helper(curr, total, index + 1)

        
        helper([], 0, 0)
       
        return res

File: Data Structures & Algorithms/target-sum/submission-2.java
class Solution {
    private int[] nums;
    private int target;

    private int res = 0;

    public int findTargetSumWays(int[] nums, int target) {
        this.nums = nums;
        this.target = target;

        helper(0, 0, true);
        helper(0, 0, false);
        return res / 2;
    }

    private void helper(int index, int current, boolean add) {
        if (index == nums.length) {
            if (current == target) res += 1;
            return;
        }
        
        int updated_current = add ? current + nums[index] : current - nums[index];
        helper(index + 1, updated_current, true);
        helper(index + 1, updated_current, false);
    }
}
